### DISCUSSION TITLE
tasmota 32 template GPIO map

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19000] [General]
Is there a map from position in GPIO array in template definition to GPIO numbers for tasmota 32?

Thank you

**Comment by sfromis** [comment]
Not outside of the code. What I'd recommend would be to create/update the template using "Configure Template", and then use the template string created when you check "Configure Other", or the `Template` command.

**Reply by sfromis** [reply]
For a "classic" ESP32 (not S2, S3 or C3), the template array and gpio numbers are like this, which is really not practical for manually positioning the pins:
```
Gpio.00     IO  160 Switch1
Gpio.01     TX    1 User
Gpio.02     IO    1 User
Gpio.03     RX    1 User
Gpio.04     IO    1 User
Gpio.05     IO    1 User
Gpio.09     FL    1 User
Gpio.10     FL    1 User
Gpio.12     IO    1 User
Gpio.13     IO    1 User
Gpio.14     IO    1 User
Gpio.15     IO    1 User
Gpio.16     IO    1 User
Gpio.17     IO    1 User
Gpio.18     IO    1 User
Gpio.19     IO    1 User
Gpio.20     IO    0 None
Gpio.21     IO  640 I2C SDA1
Gpio.22     IO  608 I2C SCL1
Gpio.23     IO    1 User
Gpio.24     IO    0 None
Gpio.25     IO    1 User
Gpio.26     IO    1 User
Gpio.27     IO    1 User
Gpio.06     FL    0 None
Gpio.07     FL    0 None
Gpio.08     FL    0 None
Gpio.11     FL    0 None
Gpio.32     AO    1 User
Gpio.33     AO    1 User
Gpio.34     IA    1 User
Gpio.35     IA    1 User
Gpio.36     IA    1 User
Gpio.37     IA    0 None
Gpio.38     IA    0 None
Gpio.39     IA    1 User
```

**Reply by msetina** [reply]
Thank you. Was in your answer an implication that it can be different for other types?
If I want to prepare template command the order for the GPIO list is needed.

**Reply by sfromis** [reply]
Yes, it is different depending on type. To prepare template commands, do not worry about the array, it works better to do as I already suggested.

**Reply by msetina** [reply]
Thank you again. I have found map for 8266 in documentation. Now I would like to extend to ESP32. If someone will be looking for this, here it is:
[0,1,2,3,4,5,9,10,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,6,7,8,11,32,33,34,35,36,37,38,39]

Is there a description for second column IO types?

**Reply by sfromis** [reply]
Doing it by a mapping of the array is really not an approach to be recommended, when you have an error-free way via the Tasmota user interface. Way too high risk of "minor mishaps".

If you mean the column with IO/FL/IA, I've not seen a full description of these. They are only used in a "reminder" column on "Configure Module", with the "A" being analog-capable pins, "FL" typically reserved for flashing, but people "should" learn the differences between pins from Espressif docs, as the nuances are not fully reflected.
